### PR TITLE
Refactor remote pinentry

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

This make Pinentry a remote component.

Some subtleties: since we reply to the original notification, we have to keep around the response object until we have a value from the user. This is kinda lame, but it's how works now, so we have a map from sessionID to response object.

A couple new things here:

remote components can take functions as props (they always could, but now we're using it). We need it to reply to the correct response object.

Actions are somewhat strongly typed! So in the pinentry reducers, the actions are typed for Pinentry actions. So if you try to access something in the payload that doesn't exist it should error.

### CODE REVIEW

File: react-native/react/actions/pinentry.js [hunk 1]
```diff
@@ -1,22 +1,96 @@
...
 import * as Constants from '../constants/pinentry'
+import engine from '../engine'
+
+import type {GUIEntryFeatures, GUIEntryArg} from '../constants/types/flow-types'
+import type {NewPinentryAction, RegisterPinentryListenerAction} from '../constants/pinentry'
+
+import type {Dispatch} from '../constants/types/flux'
+
+// TODO: there has to be a better way.
+const uglySessionIDResponseMapper: {[key: number]: Function} = {}
```
Comment (chrisnojima) [inline react-native/react/actions/pinentry.js]: We could theoretically avoid this by just keeping this within the context of the onSubmit function / onCancel function we pass to the smart component

File: react-native/react/reducers/pinentry.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  pinentryStates: {
+    [key: number]: PinentryState
+  }
+} | {
+  started: 0
+}
+
+type EnabledFeatures = {[key: string]: Feature}
+
+const intialState: RootPinentryState = {
```
Comment (cjb) [inline react-native/react/reducers/pinentry.js]: initial

File: react-native/react/reducers/pinentry.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        const features = action.payload.features
+        // Long form function to add annotation to help flow
+        const reducer = function (m, f): EnabledFeatures {
+          return {...m, f: features[f]}
+        }
+        const enabledFeatures = Object.keys(features).filter(f => features[f].allow).reduce(reducer, ({}: EnabledFeatures))
+
+        const newPinentryState: PinentryState = {
+          closed: false,
+          cancelled: false,
```
Comment (cjb) [inline react-native/react/reducers/pinentry.js]: nittiest comment ever: we use "canceled" everywhere else, technically "cancelled" is British/Canadian

Comment (MarcoPolo) [inline react-native/react/reducers/pinentry.js]: :gb: :maple_leaf: :heart:

**Comment by chrisnojima** [comment]
:+1: % cjb comments

**Pull request merged**